sources/Cdn/Controllers/CdnController.cs

```diff
@@ -0,0 +1,101 @@
+﻿using System;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Database;
+using Database.Models.Entities;
+using Grpc.Core;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.Logging;
+using FileSystem = System.IO.File;
+using Shared;
+
+namespace Cdn.Controllers
+{
+	[ApiController]
+	[Route("[controller]")]
+	public class CdnController : ControllerBase
+	{
+		private const int Seconds = 60;
+		private const int Minutes = 60;
+		private readonly string BaseAssetLocation;
+		private readonly ILogger<CdnController> _logger;
+
+		public CdnController(ILogger<CdnController> logger)
+		{
+			_logger = logger;
+			BaseAssetLocation = Environment.GetEnvironmentVariable("BASE_ASSET_LOCATION") 
+								?? throw new EnvironmentVariableMissingException("BASE_ASSET_LOCATION");
+		}
+		
+		[HttpGet("{id}")]
+		[ResponseCache(Duration = Seconds * Minutes, Location = ResponseCacheLocation.Client, NoStore = false)]
+		public async Task<IActionResult> Get(string name)
+		{
+			var headers = Request.GetTypedHeaders();
+			headers.Date = DateTimeOffset.Now;
+
+			using var context = new ArkadiaDbContext();
+
+			var cdnEntry = await context.CdnEntries.FirstAsync(asset => asset.Name == name);
+
+			if (cdnEntry is null)
+			{
+				return NotFound();
+			}
+
+			// RFC 7232 3.3 - If the content was not modified, a 304 "Not Modified" status should be sent.
+			if (!WasModified(cdnEntry))
+			{
+				// RFC 7232 4.1 - The server generating a 304 response MUST generate any of the following header fields that
+				// would have been sent in a 200 (OK) response to the same request: Cache-Control, Content-Location, Date,
+				// ETag, Expires, and Vary.
+				return NotModified();
+			}
+
+			headers.LastModified = new DateTimeOffset(cdnEntry.LastModifiedAt);
+			headers.Set("ETag", cdnEntry.ETag);
```

Comment:
Those should be added above the `File()` call (line 74), so we handle all the potentially erroneous cases first.